hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+      } catch (IOException ioe) {
+        LOG.warn("Failed to acquire client for pipeline {}, block {}",
+            pipeline, blockID);
+        throw ioe;
+      }
+    }
+  }
+
+  private synchronized int prepareRead(int len) throws IOException {
+    for (;;) {
+      if (blockPosition >= 0) {
+        if (buffersHavePosition(blockPosition)) {
+          // The current buffers have the seeked position. Adjust the buffer
+          // index and position to point to the buffer position.
+          adjustBufferPosition(blockPosition - bufferOffsetWrtBlockDataData);
+        } else {
+          // Read a required block data to fill the buffers with seeked
+          // position data
+          readDataFromContainer(len);
+        }
+      }
+      if (buffersHaveData()) {
+        // Data is available from buffers
+        ByteBuffer bb = buffers.get(bufferIndex);
+        return Math.min(len, bb.remaining());
+      } else if (dataRemainingInBlock()) {
+        // There is more data in the block stream which has not
+        // been read into the buffers yet.
+        readDataFromContainer(len);
+      } else {
+        // All available input from this block stream has been consumed.
+        return EOF;
+      }
+    }
+
+
+  }
+
+  private boolean buffersHavePosition(long pos) {
+    // Check if buffers have been allocated
+    if (buffersAllocated()) {
+      // Check if the current buffers cover the input position
+      // Released buffers should not be considered when checking if position
+      // is available
+      return pos >= bufferOffsetWrtBlockDataData +
+          bufferoffsets.get(0) &&
+          pos < bufferOffsetWrtBlockDataData + buffersSize;
+    }
+    return false;
+  }
+
+  /**
+   * Check if the buffers have been allocated data and false otherwise.
+   */
+  @VisibleForTesting
+  protected boolean buffersAllocated() {
+    return buffers != null && !buffers.isEmpty();
+  }
+
+  /**
+   * Adjust the buffers position to account for seeked position and/ or checksum
+   * boundary reads.
+   * @param bufferPosition the position to which the buffers must be advanced
+   */
+  private void adjustBufferPosition(long bufferPosition) {
+    // The bufferPosition is w.r.t the current buffers.
+    // Adjust the bufferIndex and position to the seeked bufferPosition.
+    if (bufferIndex >= buffers.size()) {
+      bufferIndex = Collections.binarySearch(bufferoffsets, bufferPosition);
+    } else if (bufferPosition < bufferoffsets.get(bufferIndex)) {
+      bufferIndex = Collections.binarySearch(
+          bufferoffsets.subList(0, bufferIndex), bufferPosition);
+    } else if (bufferPosition >= bufferoffsets.get(bufferIndex) +
+        buffers.get(bufferIndex).capacity()) {
+      bufferIndex = Collections.binarySearch(bufferoffsets.subList(
+          bufferIndex + 1, buffers.size()), bufferPosition);
+    }
+    if (bufferIndex < 0) {
+      bufferIndex = -bufferIndex - 2;
+    }
```

Comment:
Can we add comment to clarify why we are subtracting 2?

Reply:
done